packages/react-dom/src/events/ReactDOMEventReplaying.js

```diff
@@ -615,3 +615,41 @@ export function retryIfBlockedOn(
     }
   }
 }
+
+function replayEvent(
+  queuedEvent: QueuedReplayableEvent,
+  targetContainer: EventTarget,
+) {
+  if (enableCapturePhaseSelectiveHydrationWithoutDiscreteEventReplay) {
+    nativeDispatchForReplay(queuedEvent.nativeEvent, targetContainer);
+  } else {
+    setEventIsReplaying(queuedEvent.nativeEvent);
+    dispatchEventForPluginEventSystem(
+      queuedEvent.domEventName,
+      queuedEvent.eventSystemFlags,
+      queuedEvent.nativeEvent,
+      getClosestInstanceFromNode(getEventTarget(queuedEvent.nativeEvent)),
+      targetContainer,
+    );
+  }
+}
+
+export const nativeDispatchForReplay = (
+  event: AnyNativeEvent,
+  targetContainer: EventTarget,
+) => {
+  /**
+   * When we queue continuous events we queue them in both capture and bubble phase
+   * but We only want to dispatch them once though. Some events only have a bubble phase
+   * so gating to queue only in capture phase would cause us to miss those events.
+   * Instead we queue both and use this WeakSet to make sure we don't double dispatch events.
+   * This has the benefit of not needing to hook into the native events stopPropagation() since
+   * the bubble phase will not be fired by the browser if the event has stopPropagation called on it.
+   */
+  if (!isReplayingEvent(event)) {
+    const eventClone = new event.constructor(event.type, (event: any));
+    setEventIsReplaying(event);
+    setEventIsReplaying(eventClone);
```

Comment:
This is making me a bit queasy. My instinct might be off but it feels like we should have enough information to thread `isReplaying` through somehow instead of "guessing" whether we're replaying or not by "tagging" events. E.g. if we want to avoid duplicates while queueing, can we make a check *while queueing* based on what's in the queue? If it weren't possible to use expandos, would there still be a solution, or is there some constraint that makes this complexity unavoidable?

Reply:
The concept I'm concerned about is that this introduces extra bit of statefulness to event objects. They go from "not replaying" to "replaying" and forever staying there. Previously, there was no statefulness because each time we replayed, we _knew_ we're replaying. Is there any way we can keep that property?